store/src/db/migrations.rs

```diff
@@ -1,71 +1,90 @@
-            CONSTRAINT nullifiers_nullifier_prefix_is_u16 CHECK (nullifier_prefix BETWEEN 0 AND 0xFFFF),
-            CONSTRAINT nullifiers_block_num_is_u32 CHECK (block_num BETWEEN 0 AND 0xFFFFFFFF)
-        ) STRICT, WITHOUT ROWID;
-        ",
-    )])
-});
+use rusqlite::Connection;
+use rusqlite_migration::{Migrations, SchemaVersion, M};
+use tracing::{debug, info, instrument};
+
+use crate::{db::settings::Settings, errors::DatabaseError, COMPONENT};
+
+type Hash = Blake3Digest<20>;
+
+const MIGRATION_SCRIPTS: [&str; 1] = [include_str!("migrations/001-init.sql")];
+static MIGRATION_HASHES: Lazy<Vec<Hash>> = Lazy::new(compute_migration_hashes);
+static MIGRATIONS: Lazy<Migrations> = Lazy::new(prepare_migrations);
+
+fn up(s: &'static str) -> M<'static> {
+    M::up(s).foreign_key_check()
+}
+
+const DB_MIGRATION_HASH_FIELD: &str = "db-migration-hash";
+
+#[instrument(target = "miden-store", skip_all, err)]
+pub fn apply_migrations(conn: &mut Connection) -> super::Result<()> {
+    let version_before = MIGRATIONS.current_version(conn)?;
+
+    info!(target: COMPONENT, version_before = %version_before, "Running database migrations");
+
+    if let SchemaVersion::Inside(ver) = version_before {
+        if !Settings::exists(conn)? {
+            return Err(DatabaseError::UnsupportedDatabaseVersion);
+        }
+
+        let expected_hash = MIGRATION_HASHES[ver.get() - 1].to_bytes();
+        let actual_hash = Settings::get_value(conn, DB_MIGRATION_HASH_FIELD)?;
+
+        debug!(
+            target: COMPONENT,
+            expected_hash = %hex::encode(&expected_hash),
+            actual_hash = ?actual_hash.as_ref().map(hex::encode),
+            "Comparing migration hashes",
+        );
+
+        if actual_hash != Some(expected_hash) {
+            return Err(DatabaseError::UnsupportedDatabaseVersion);
+        }
+    }
+
+    MIGRATIONS.to_latest(conn).map_err(DatabaseError::MigrationError)?;
+
+    if version_before != MIGRATIONS.current_version(conn)? {
+        let last_hash = MIGRATION_HASHES[MIGRATION_HASHES.len() - 1].to_bytes();
+        debug!(target: COMPONENT, new_hash = %hex::encode(&last_hash), "Updating migration hash in settings table");
+        Settings::set_value(conn, DB_MIGRATION_HASH_FIELD, &last_hash)?;
+    }
+
+    Ok(())
+}
+
+fn prepare_migrations() -> Migrations<'static> {
+    Migrations::new(MIGRATION_SCRIPTS.map(up).to_vec())
+}
+
+fn compute_migration_hashes() -> Vec<Hash> {
+    let mut accumulator = Hash::default();
+    MIGRATION_SCRIPTS
+        .iter()
+        .map(|sql| {
+            let script_hash = Blake3_160::hash(preprocess_sql(sql).as_bytes());
+            accumulator = Blake3_160::merge(&[accumulator, script_hash]);
+            accumulator
+        })
+        .collect()
+}
+
+fn preprocess_sql(sql: &str) -> String {
+    // TODO: We can also remove all comments here (need to analyze the SQL script in order to remain comments
+    //       in string literals).
+    remove_spaces(sql)
```

Comment:
This is a tricky one, one possibility is to use `PRAGMA table_info(?table)` to get the parsed table, but that would required creating a new DB, applying all the transactions, then getting the results of `table_info` from the fresh DB and comparing against the "online" DB.

I'm not sure how much work it is, so I suppose this is the best approach for now (in terms of effort to get it working).